File: code/packages/python/blas-library/tests/test_cross_backend.py
"""Cross-backend equivalence tests — same computation through all 7 backends.

=== The Capstone Test ===

This is the most important test file in the BLAS library. It runs the
same BLAS operations through ALL 7 backends and verifies they produce
identical results (within floating-point tolerance).

If any backend disagrees with the others, the bug is in that backend.
The CPU backend is the reference — all GPU backends must match it.
"""

from __future__ import annotations

import pytest
from conftest import approx_equal, approx_list

from blas_library import (
    Matrix,
    Side,
    Transpose,
    Vector,
    create_blas,
)

# All 7 backend names
ALL_BACKENDS = ["cpu", "cuda", "opencl", "metal", "vulkan", "webgpu", "opengl"]


@pytest.fixture(params=ALL_BACKENDS)
def backend_name(request: pytest.FixtureRequest) -> str:
    """Parametrize over all 7 backend names."""
    return request.param


@pytest.fixture
def blas(backend_name: str) -> object:
    """Create a BLAS instance for the given backend."""
    return create_blas(backend_name)


# =========================================================================
# Cross-backend Level 1 tests
# =========================================================================


class TestCrossBackendLevel1:
    """All backends must agree on Level 1 operations."""

    def test_saxpy(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, 2.0, 3.0, 4.0], size=4)
        y = Vector(data=[5.0, 6.0, 7.0, 8.0], size=4)
        result = blas.saxpy(2.0, x, y)  # type: ignore[union-attr]
        expected = [7.0, 10.0, 13.0, 16.0]
        assert approx_list(result.data, expected), f"{backend_name} SAXPY failed"

    def test_sdot(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, 2.0, 3.0], size=3)
        y = Vector(data=[4.0, 5.0, 6.0], size=3)
        result = blas.sdot(x, y)  # type: ignore[union-attr]
        assert approx_equal(result, 32.0), f"{backend_name} DOT failed"

    def test_snrm2(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[3.0, 4.0], size=2)
        result = blas.snrm2(x)  # type: ignore[union-attr]
        assert approx_equal(result, 5.0), f"{backend_name} NRM2 failed"

    def test_sscal(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, 2.0, 3.0], size=3)
        result = blas.sscal(3.0, x)  # type: ignore[union-attr]
        assert approx_list(result.data, [3.0, 6.0, 9.0]), f"{backend_name} SCAL failed"

    def test_sasum(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, -2.0, 3.0, -4.0], size=4)
        result = blas.sasum(x)  # type: ignore[union-attr]
        assert approx_equal(result, 10.0), f"{backend_name} ASUM failed"

    def test_isamax(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, -5.0, 3.0], size=3)
        result = blas.isamax(x)  # type: ignore[union-attr]
        assert result == 1, f"{backend_name} IAMAX failed"

    def test_scopy(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, 2.0, 3.0], size=3)
        result = blas.scopy(x)  # type: ignore[union-attr]
        assert approx_list(result.data, [1.0, 2.0, 3.0]), f"{backend_name} COPY failed"

    def test_sswap(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, 2.0], size=2)
        y = Vector(data=[3.0, 4.0], size=2)
        new_x, new_y = blas.sswap(x, y)  # type: ignore[union-attr]
        assert approx_list(new_x.data, [3.0, 4.0]), f"{backend_name} SWAP-x failed"
        assert approx_list(new_y.data, [1.0, 2.0]), f"{backend_name} SWAP-y failed"


# =========================================================================
# Cross-backend Level 2 tests
# =========================================================================


class TestCrossBackendLevel2:
    """All backends must agree on Level 2 operations."""

    def test_sgemv_no_trans(self, blas: object, backend_name: str) -> None:
        a = Matrix(data=[1.0, 2.0, 3.0, 4.0, 5.0, 6.0], rows=2, cols=3)
        x = Vector(data=[1.0, 1.0, 1.0], size=3)
        y = Vector(data=[0.0, 0.0], size=2)
        result = blas.sgemv(Transpose.NO_TRANS, 1.0, a, x, 0.0, y)  # type: ignore[union-attr]
        assert approx_list(result.data, [6.0, 15.0]), f"{backend_name} GEMV failed"

    def test_sger(self, blas: object, backend_name: str) -> None:
        x = Vector(data=[1.0, 2.0], size=2)
        y = Vector(data=[3.0, 4.0], size=2)
        a = Matrix(data=[0.0] * 4, rows=2, cols=2)
        result = blas.sger(1.0, x, y, a)  # type: ignore[union-attr]
        expected = [3.0, 4.0, 6.0, 8.0]
        assert approx_list(result.data, expected), f"{backend_name} GER failed"


# =========================================================================
# Cross-backend Level 3 tests — THE BIG ONE
# =========================================================================


class TestCrossBackendLevel3:
    """All backends must agree on Level 3 operations."""

    def test_sgemm_canonical(self, blas: object, backend_name: str) -> None:
        """The canonical GEMM test from the spec:
        A = [[1,2,3],[4,5,6]]  B = [[7,8],[9,10],[11,12]]
        C = A*B = [[58,64],[139,154]]
        """
        a = Matrix(data=[1.0, 2.0, 3.0, 4.0, 5.0, 6.0], rows=2, cols=3)
        b = Matrix(data=[7.0, 8.0, 9.0, 10.0, 11.0, 12.0], rows=3, cols=2)
        c = Matrix(data=[0.0] * 4, rows=2, cols=2)
        result = blas.sgemm(  # type: ignore[union-attr]
            Transpose.NO_TRANS, Transpose.NO_TRANS, 1.0, a, b, 0.0, c
        )
        expected = [58.0, 64.0, 139.0, 154.0]
        assert approx_list(result.data, expected), (
            f"{backend_name} GEMM failed: got {result.data}"
        )

    def test_sgemm_with_alpha_beta(self, blas: object, backend_name: str) -> None:
        """GEMM with alpha=2, beta=3."""
        a = Matrix(data=[1.0, 0.0, 0.0, 1.0], rows=2, cols=2)
        b = Matrix(data=[5.0, 6.0, 7.0, 8.0], rows=2, cols=2)
        c = Matrix(data=[1.0, 1.0, 1.0, 1.0], rows=2, cols=2)
        result = blas.sgemm(  # type: ignore[union-attr]
            Transpose.NO_TRANS, Transpose.NO_TRANS, 2.0, a, b, 3.0, c
        )
        expected = [13.0, 15.0, 17.0, 19.0]
        assert approx_list(result.data, expected), (
            f"{backend_name} GEMM alpha/beta failed"
        )

    def test_ssymm(self, blas: object, backend_name: str) -> None:
        a = Matrix(data=[1.0, 2.0, 2.0, 1.0], rows=2, cols=2)
        b = Matrix(data=[1.0, 0.0, 0.0, 1.0], rows=2, cols=2)
        c = Matrix(data=[0.0] * 4, rows=2, cols=2)
        result = blas.ssymm(Side.LEFT, 1.0, a, b, 0.0, c)  # type: ignore[union-attr]
        assert approx_list(result.data, [1.0, 2.0, 2.0, 1.0]), (
            f"{backend_name} SYMM failed"
        )

    def test_sgemm_batched(self, blas: object, backend_name: str) -> None:
        a1 = Matrix(data=[2.0, 0.0, 0.0, 2.0], rows=2, cols=2)
        b1 = Matrix(data=[1.0, 2.0, 3.0, 4.0], rows=2, cols=2)
        c1 = Matrix(data=[0.0] * 4, rows=2, cols=2)
        results = blas.sgemm_batched(  # type: ignore[union-attr]
            Transpose.NO_TRANS,
            Transpose.NO_TRANS,
            1.0,
            [a1],
            [b1],
            0.0,
            [c1],
        )
        assert approx_list(results[0].data, [2.0, 4.0, 6.0, 8.0]), (
            f"{backend_name} batched GEMM failed"
        )

File: code/packages/rust/blas-library/tests/test_cross_backend.rs
//! Cross-backend consistency tests: verify all 7 backends produce identical results.
//!
//! These tests run the same operation on every backend and assert that the
//! results match the CPU reference exactly. This is the core quality guarantee.

use blas_library::traits::BlasBackend;
use blas_library::{BackendRegistry, Matrix, Side, Transpose, Vector};

fn all_backends() -> Vec<(String, Box<dyn BlasBackend>)> {
    let registry = BackendRegistry::with_defaults();
    registry
        .list_available()
        .into_iter()
        .map(|name| {
            let backend = registry.get(&name).unwrap();
            (name, backend)
        })
        .collect()
}

// =========================================================================
// Level 1 cross-backend tests
// =========================================================================

#[test]
fn test_cross_saxpy() {
    let x = Vector::new(vec![1.0, -2.0, 3.0, -4.0, 5.0]);
    let y = Vector::new(vec![5.0, 4.0, -3.0, 2.0, -1.0]);
    let alpha = 2.5;

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.saxpy(alpha, &x, &y).unwrap();
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SAXPY mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

#[test]
fn test_cross_sdot() {
    let x = Vector::new(vec![1.0, 2.0, 3.0, 4.0]);
    let y = Vector::new(vec![-1.0, 2.0, -3.0, 4.0]);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.sdot(&x, &y).unwrap();
        if let Some(exp) = expected {
            assert_eq!(result, exp, "SDOT mismatch on {}", name);
        } else {
            expected = Some(result);
        }
    }
}

#[test]
fn test_cross_snrm2() {
    let x = Vector::new(vec![1.0, 2.0, 3.0, 4.0, 5.0]);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.snrm2(&x);
        if let Some(exp) = expected {
            assert_eq!(result, exp, "SNRM2 mismatch on {}", name);
        } else {
            expected = Some(result);
        }
    }
}

#[test]
fn test_cross_sscal() {
    let x = Vector::new(vec![1.0, -2.0, 3.0]);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.sscal(-3.0, &x);
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SSCAL mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

#[test]
fn test_cross_sasum() {
    let x = Vector::new(vec![-1.0, 2.0, -3.0, 4.0, -5.0]);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.sasum(&x);
        if let Some(exp) = expected {
            assert_eq!(result, exp, "SASUM mismatch on {}", name);
        } else {
            expected = Some(result);
        }
    }
}

#[test]
fn test_cross_isamax() {
    let x = Vector::new(vec![1.0, -10.0, 5.0, -3.0]);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.isamax(&x);
        if let Some(exp) = expected {
            assert_eq!(result, exp, "ISAMAX mismatch on {}", name);
        } else {
            expected = Some(result);
        }
    }
}

#[test]
fn test_cross_scopy() {
    let x = Vector::new(vec![3.14, 2.71, 1.41]);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.scopy(&x);
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SCOPY mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

#[test]
fn test_cross_sswap() {
    let x = Vector::new(vec![1.0, 2.0, 3.0]);
    let y = Vector::new(vec![4.0, 5.0, 6.0]);

    let mut expected_x = None;
    let mut expected_y = None;
    for (name, backend) in all_backends() {
        let (rx, ry) = backend.sswap(&x, &y).unwrap();
        if let Some(ref ex) = expected_x {
            assert_eq!(rx.data(), ex, "SSWAP X mismatch on {}", name);
            assert_eq!(
                ry.data(),
                expected_y.as_ref().unwrap(),
                "SSWAP Y mismatch on {}",
                name
            );
        } else {
            expected_x = Some(rx.data().to_vec());
            expected_y = Some(ry.data().to_vec());
        }
    }
}

// =========================================================================
// Level 2 cross-backend tests
// =========================================================================

#[test]
fn test_cross_sgemv() {
    let a = Matrix::new(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 2, 3);
    let x = Vector::new(vec![1.0, 2.0, 3.0]);
    let y = Vector::new(vec![10.0, 20.0]);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend
            .sgemv(Transpose::NoTrans, 2.0, &a, &x, 1.0, &y)
            .unwrap();
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SGEMV mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

#[test]
fn test_cross_sger() {
    let x = Vector::new(vec![1.0, 2.0, 3.0]);
    let y = Vector::new(vec![4.0, 5.0]);
    let a = Matrix::new(vec![10.0, 20.0, 30.0, 40.0, 50.0, 60.0], 3, 2);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.sger(2.0, &x, &y, &a).unwrap();
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SGER mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

// =========================================================================
// Level 3 cross-backend tests
// =========================================================================

#[test]
fn test_cross_sgemm() {
    let a = Matrix::new(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 2, 3);
    let b = Matrix::new(vec![7.0, 8.0, 9.0, 10.0, 11.0, 12.0], 3, 2);
    let c = Matrix::new(vec![1.0, 2.0, 3.0, 4.0], 2, 2);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend
            .sgemm(Transpose::NoTrans, Transpose::NoTrans, 2.0, &a, &b, 0.5, &c)
            .unwrap();
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SGEMM mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

#[test]
fn test_cross_sgemm_trans() {
    let a = Matrix::new(vec![1.0, 3.0, 2.0, 4.0], 2, 2);
    let b = Matrix::new(vec![5.0, 6.0, 7.0, 8.0], 2, 2);
    let c = Matrix::zeros(2, 2);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend
            .sgemm(Transpose::Trans, Transpose::NoTrans, 1.0, &a, &b, 0.0, &c)
            .unwrap();
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SGEMM Trans mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

#[test]
fn test_cross_ssymm() {
    let a = Matrix::new(vec![2.0, 1.0, 1.0, 3.0], 2, 2);
    let b = Matrix::new(vec![1.0, 2.0, 3.0, 4.0], 2, 2);
    let c = Matrix::zeros(2, 2);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend.ssymm(Side::Left, 1.0, &a, &b, 0.0, &c).unwrap();
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SSYMM mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}

#[test]
fn test_cross_sgemm_batched() {
    let a_list = vec![
        Matrix::new(vec![1.0, 2.0, 3.0, 4.0], 2, 2),
        Matrix::new(vec![5.0, 6.0, 7.0, 8.0], 2, 2),
    ];
    let b_list = vec![
        Matrix::new(vec![1.0, 0.0, 0.0, 1.0], 2, 2),
        Matrix::new(vec![2.0, 0.0, 0.0, 2.0], 2, 2),
    ];
    let c_list = vec![Matrix::zeros(2, 2), Matrix::zeros(2, 2)];

    let mut expected: Option<Vec<Vec<f32>>> = None;
    for (name, backend) in all_backends() {
        let results = backend
            .sgemm_batched(
                Transpose::NoTrans,
                Transpose::NoTrans,
                1.0,
                &a_list,
                &b_list,
                0.0,
                &c_list,
            )
            .unwrap();
        if let Some(ref exp) = expected {
            for (i, r) in results.iter().enumerate() {
                assert_eq!(
                    r.data(),
                    exp[i].as_slice(),
                    "SGEMM_BATCHED[{}] mismatch on {}",
                    i,
                    name
                );
            }
        } else {
            expected = Some(results.iter().map(|r| r.data().to_vec()).collect::<Vec<_>>());
        }
    }
}

// =========================================================================
// Additional cross-backend tests for error cases
// =========================================================================

#[test]
fn test_cross_saxpy_error() {
    let x = Vector::new(vec![1.0]);
    let y = Vector::new(vec![1.0, 2.0]);
    for (name, backend) in all_backends() {
        assert!(
            backend.saxpy(1.0, &x, &y).is_err(),
            "{} should reject mismatched SAXPY",
            name
        );
    }
}

#[test]
fn test_cross_sdot_error() {
    let x = Vector::new(vec![1.0, 2.0]);
    let y = Vector::new(vec![1.0]);
    for (name, backend) in all_backends() {
        assert!(
            backend.sdot(&x, &y).is_err(),
            "{} should reject mismatched SDOT",
            name
        );
    }
}

#[test]
fn test_cross_sswap_error() {
    let x = Vector::new(vec![1.0]);
    let y = Vector::new(vec![1.0, 2.0]);
    for (name, backend) in all_backends() {
        assert!(
            backend.sswap(&x, &y).is_err(),
            "{} should reject mismatched SSWAP",
            name
        );
    }
}

#[test]
fn test_cross_sgemm_error() {
    let a = Matrix::new(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 2, 3);
    let b = Matrix::new(vec![1.0, 2.0, 3.0, 4.0], 2, 2);
    let c = Matrix::zeros(2, 2);
    for (name, backend) in all_backends() {
        assert!(
            backend
                .sgemm(Transpose::NoTrans, Transpose::NoTrans, 1.0, &a, &b, 0.0, &c)
                .is_err(),
            "{} should reject mismatched SGEMM",
            name
        );
    }
}

#[test]
fn test_cross_sgemv_error() {
    let a = Matrix::new(vec![1.0, 2.0, 3.0, 4.0], 2, 2);
    let x = Vector::new(vec![1.0, 2.0, 3.0]);
    let y = Vector::zeros(2);
    for (name, backend) in all_backends() {
        assert!(
            backend
                .sgemv(Transpose::NoTrans, 1.0, &a, &x, 0.0, &y)
                .is_err(),
            "{} should reject mismatched SGEMV",
            name
        );
    }
}

#[test]
fn test_cross_sger_error() {
    let x = Vector::new(vec![1.0, 2.0, 3.0]);
    let y = Vector::new(vec![1.0, 2.0]);
    let a = Matrix::zeros(2, 2);
    for (name, backend) in all_backends() {
        assert!(
            backend.sger(1.0, &x, &y, &a).is_err(),
            "{} should reject mismatched SGER",
            name
        );
    }
}

#[test]
fn test_cross_ssymm_not_square() {
    let a = Matrix::new(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 2, 3);
    let b = Matrix::zeros(2, 2);
    let c = Matrix::zeros(2, 2);
    for (name, backend) in all_backends() {
        assert!(
            backend.ssymm(Side::Left, 1.0, &a, &b, 0.0, &c).is_err(),
            "{} should reject non-square SSYMM",
            name
        );
    }
}

#[test]
fn test_cross_sgemm_batched_error() {
    let a = vec![Matrix::zeros(2, 2)];
    let b = vec![Matrix::zeros(2, 2), Matrix::zeros(2, 2)];
    let c = vec![Matrix::zeros(2, 2)];
    for (name, backend) in all_backends() {
        assert!(
            backend
                .sgemm_batched(Transpose::NoTrans, Transpose::NoTrans, 1.0, &a, &b, 0.0, &c)
                .is_err(),
            "{} should reject mismatched batched GEMM sizes",
            name
        );
    }
}

#[test]
fn test_cross_sgemv_transposed() {
    let a = Matrix::new(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 2, 3);
    let x = Vector::new(vec![1.0, 1.0]);
    let y = Vector::zeros(3);

    let mut expected = None;
    for (name, backend) in all_backends() {
        let result = backend
            .sgemv(Transpose::Trans, 1.0, &a, &x, 0.0, &y)
            .unwrap();
        if let Some(ref exp) = expected {
            assert_eq!(result.data(), exp, "SGEMV Trans mismatch on {}", name);
        } else {
            expected = Some(result.data().to_vec());
        }
    }
}
